indexing-service/src/main/java/io/druid/indexing/overlord/RemoteTaskRunner.java

```diff
@@ -619,40 +631,54 @@ private boolean tryAssignTask(final Task task, final RemoteTaskRunnerWorkItem ta
-          ImmutableMap.copyOf(
-              Maps.transformEntries(
-                  Maps.filterEntries(
-                      zkWorkers, new Predicate<Map.Entry<String, ZkWorker>>()
-                      {
-                        @Override
-                        public boolean apply(Map.Entry<String, ZkWorker> input)
+
+      ZkWorker assignedWorker = null;
+      try {
+        final Optional<ImmutableZkWorker> immutableZkWorker = strategy.findWorkerForTask(
+            config,
+            ImmutableMap.copyOf(
+                Maps.transformEntries(
+                    Maps.filterEntries(
+                        zkWorkers, new Predicate<Map.Entry<String, ZkWorker>>()
                         {
-                          return !lazyWorkers.containsKey(input.getKey());
+                          @Override
+                          public boolean apply(Map.Entry<String, ZkWorker> input)
+                          {
+                            return !lazyWorkers.containsKey(input.getKey()) &&
+                                   !workersWithUnacknowledgedTask.containsKey(input.getKey());
+                          }
                         }
-                      }
-                  ),
-                  new Maps.EntryTransformer<String, ZkWorker, ImmutableZkWorker>()
-                  {
-                    @Override
-                    public ImmutableZkWorker transformEntry(
-                        String key, ZkWorker value
-                    )
+                    ),
+                    new Maps.EntryTransformer<String, ZkWorker, ImmutableZkWorker>()
                     {
-                      return value.toImmutable();
+                      @Override
+                      public ImmutableZkWorker transformEntry(
+                          String key, ZkWorker value
+                      )
+                      {
+                        return value.toImmutable();
+                      }
                     }
-                  }
-              )
-          ),
-          task
-      );
-      if (immutableZkWorker.isPresent()) {
-        final ZkWorker zkWorker = zkWorkers.get(immutableZkWorker.get().getWorker().getHost());
-        return announceTask(task, zkWorker, taskRunnerWorkItem);
-      } else {
+                )
+            ),
+            task
+        );
+
+        if (immutableZkWorker.isPresent()
+            &&
+            workersWithUnacknowledgedTask.putIfAbsent(immutableZkWorker.get().getWorker().getHost(), task.getId())
+            == null) {
+          assignedWorker = zkWorkers.get(immutableZkWorker.get().getWorker().getHost());
+          return announceTask(task, assignedWorker, taskRunnerWorkItem);
+        }
+
         log.debug("Worker nodes %s do not have capacity to run any more tasks!", zkWorkers.values());
         return false;
+      } finally {
+        if (assignedWorker != null) {
+          workersWithUnacknowledgedTask.remove(assignedWorker.getWorker().getHost());
```

Comment:
I believe this shouldn't run if `workersWithUnacknowledgedTask.putIfAbsent` had returned null